### ISSUE TITLE
Comment considered harmful

### ISSUE THREAD

**Issue by sorawee** [bootstrapworld/codemirror-blocks#146]
Consider:

```
hello ; this is a comment
world
```

We can add a node after `world` without any problem. However, if we add a node after the `hello` block, the node won't become a block.

![2018-08-15-151354_322x179_scrot](https://user-images.githubusercontent.com/9099577/44138360-ddce93a6-a09d-11e8-84ce-9ea10f8ffe3d.png)

Worse, after switching back to text mode, here's the appearance:

![2018-08-15-151617_274x114_scrot](https://user-images.githubusercontent.com/9099577/44138452-3505d954-a09e-11e8-95f4-124263483cf8.png)

But if one copy and paste the whole code to another editor, we will see:

```
hello ; this is a comment (a b)
world (a b)
```

So `; this is a comment` is somehow not restored properly to the text mode.

Switching back to the block mode one more time. Now we have:

![2018-08-15-151826_319x199_scrot](https://user-images.githubusercontent.com/9099577/44138555-849ef7ac-a09e-11e8-9237-b7e94a5040f1.png)

where `(a b)` becomes the inline comment of the block `hello`.

**Comment by schanzer** [comment]
1) Yup - we need a better solution for dealing with line comments (more below)
2) I can't reproduce the "deleted comment" you're seeing in FF or Chrome. Instead, I get the text inserted node...after the comment (which is still broken, but what you'd expect. Are you _typing_ that new block, or dragging it in from elsewhere?

As for dealing with comments:
My original plan for the block environment didn't include comments at all (a side-effect of the fact that the WeScheme parser just stripped out comments from the AST to begin with, and therefore we didn't _have_ comments!). Since they were never formalized, a real description of how they operate was never added to the spec. But obviously yeah: we need comments, dammit.

...but clearly it's not as simple as just letting the parser return them. For now, comments are handled in a slapdash way so that they can at least be read aloud. What's really needed, however, is:
1. Comments should be able to exist anywhere in the AST, and be associated with 0-1 other ASTNodes
2. That means some comments should be globally visible (e.g. a comment at the top level that is not associated with any node)
2. All comment should be converted to block comments when going from blocks->text
3. Comments should be editable (via a hotkey, or double-clicking on a visible comment), with the enclosing block comment characters added automatically

If this behavior sounds right to you and @justinpombrio , I'll write it into the spec and create a Git issue for it.

**Comment by sorawee** [comment]
I was typing a new block.

Note that I am using `editor.html` which has a proper text editing mode in CM. `wescheme.html` simply destroys CM to get a plain textarea and construct CM again when it's needed, so I expect to see no problem there.

**Comment by sorawee** [comment]
Re 1 and 2:

I want to know how do you envision these turning into blocks:

```
(define (f x)
  ; def



  ; abc
  x)
```

```
(define (f x)
  ; abc
  x
  ; def
   )
```

> 3. All comment should be converted to block comments when going from blocks->text

Does that mean `block->text(text->block(p))` might not equal `p`? It's OK but might be considered an undesirable property.

4. Sounds right to me.

**Comment by schanzer** [comment]
The first example should result in a block identical to `(define (f x)  x)`, with the comment `abc` associated with `x`. Not sure about `def` yet -- one option is to combine it with `abc`, but I'm open to other suggestions. The second should result in the same block, with `blah` associated with `x`. Same issues for the remaining comments -- I'd be curious to know what you think. Perhaps what's needed is a comment node that's treated differently.

As for 3: yes, `block->text(text->block(p))` would not necessarily be the identity. But that was already going to be the case _somewhat_, since Justin's work may change the whitespace. I also think this is a reasonable compromise to make.

**Comment by sorawee** [comment]
So `; abc` in one context is associated to `x` and it's not in another?

**Comment by schanzer** [comment]
It depends on how the `comment<->node` association heuristic works. In Racket, the heuristic I used is that a comment immediately to the right of a node is associated with that node, and if there is none then the comment on the previous line is associated with it. This happens at the parser level, so different languages can use different heuristics.

**Comment by justinpombrio** [comment]
@schanzer I agree with everything you've said. There's some trickiness in that languages' parsers typically ignore comments, but it's _already_ the case that the parser needs to be extended a bit for e.g. blanks, and it seems entirely reasonable to say that if you want the block editor to support comments, your parser needs to include comments.

And yes, `text->blocks->text` is not the identity. I expect the other direction `blocks->text->blocks` to be the identity, though.

**Comment by schanzer** [comment]
100% agree on the last point.

**Comment by schanzer** [comment]
@justinpombrio track this issue and confirm it's addressed before merging your work into `new-reactify`

**Comment by schanzer** [comment]
Justin, when we spoke about this before we’d agreed to allow one line comment to the right of some code. What changed?

**Comment by justinpombrio** [comment]
Ah, I thought this had come across in my email about pretty-printing comments.

To be clear, the behavior right now is that you can write comments wherever you want (including deeply nested inside code), but line-comments at the end of a line get pretty-printed to be on the line before. The block view is identical in both cases.

So you'd like the line comments to stay as line comments, and not get reformatted, right?

Right now there's no good way to tell whether a comment is a line comment at the end of a line or not, AFAIK. Either the parser could keep track of it, or we could do something hacky by trying to see if it's the only thing on the line or not. Once we can tell whether a comment is a line-comment, fixing this is just a matter of putting an `if` statement in the pretty-printing function that formats the comment one way or the other.

**Comment by justinpombrio** [comment]
We now convert all comments into block (`#| ... |#`) comments when pretty-printing.

**Comment by justinpombrio** [comment]
I believe this issue is now fixed, except for the "Lost Comment" issue #180. That looks like a WeScheme parsing issue, so @schanzer is looking into it.

Modulo that issue, I believe you can now drag&drop and cut/copy/paste commented blocks from anywhere to anywhere else safely, and with the comment staying on its original block.
